File: 153.find-minimum-in-rotated-sorted-array.java
/*
 * @lc app=leetcode id=153 lang=java
 *
 * [153] Find Minimum in Rotated Sorted Array
 */

// @lc code=start

class Solution {
    /*
     * Here is an intuition for updating the search space boundaries:
     * 
     * If nums[mid] < nums[hi], nums[mid] or some value before it could be our
     * inflection point. Therefore, let hi = mid, including mid in our new search
     * space.
     * 
     * If nums[mid] > nums[hi], nums[mid] cannot be our inflection point. Candidates
     * are to the right of it. Let lo = mid + 1, exluding the mid from the search
     * space.
     */
    public int findMin(int[] nums) {
        int left = 0, right = nums.length - 1;
        while (left < right) {
            int mid = left + (right - left) / 2;
            if (nums[mid] < nums[right]) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }

        return nums[left];
    }
}
// @lc code=end

File: 153.find-minimum-in-rotated-sorted-array.py
#
# @lc app=leetcode id=153 lang=python3
#
# [153] Find Minimum in Rotated Sorted Array
#

# @lc code=start
class Solution:
    def findMin(self, nums: List[int]) -> int:
        # unique elements, so it is monotonically increasing
        n = len(nums)
        left, right = 0, n-1
        # since we need to find minimum, it is unique, we use '<='
        while left<=right:
            mid = (left+right)//2
            if nums[mid]>=nums[right]:
                left=mid+1
            else:
                right=mid

        return nums[right]
    
# class Solution:
#     def findMin(self, nums: List[int]) -> int:
#         l, r = 0, len(nums)-1

#         while l < r:
#             mid = (l+r)//2
#             # All the integers of nums are unique.
#             if nums[mid] < nums[r]:
#                 r = mid
#             else:
#                 l = mid+1

#         return nums[l]

# @lc code=end
